src/incubator/TransitionAnimator/index.tsx

```diff
@@ -0,0 +1,140 @@
+/* eslint-disable react-hooks/exhaustive-deps */
+import React, {PropsWithChildren, useEffect, useCallback, useImperativeHandle} from 'react';
+import {View as RNView, LayoutChangeEvent} from 'react-native';
+import Animated, {useSharedValue, useAnimatedStyle, withSpring, withTiming, runOnJS} from 'react-native-reanimated';
+import View, {ViewProps} from '../../components/view';
+import {forwardRef, ForwardRefInjectedProps} from '../../commons/new';
+import useHiddenLocations, {HiddenLocation} from './useHiddenLocations';
+const AnimatedView = Animated.createAnimatedComponent(View);
+export {HiddenLocation as TransitionLocation};
+
+const DEFAULT_ANIMATION_VELOCITY = 300;
+const DEFAULT_ANIMATION_CONFIG = {velocity: DEFAULT_ANIMATION_VELOCITY, damping: 18, stiffness: 300, mass: 0.4};
+
+export type TransitionAnimationEndType = 'in' | 'out';
+
+export interface TransitionAnimatorProps extends ViewProps {
+  /**
+   * Callback to the animation end.
+   */
+  onAnimationEnd?: (animationType: TransitionAnimationEndType) => void;
+  /**
+   * If this is given there will be an enter animation from this location.
+   */
+  enterAnimationLocation?: HiddenLocation;
+  /**
+   * If this is given there will be an exit animation in this location.
+   */
+  exitAnimationLocation?: HiddenLocation;
+}
+
+type Props = PropsWithChildren<TransitionAnimatorProps> & ForwardRefInjectedProps;
+interface Statics {
+  animateOut: () => void;
+}
+
+const TransitionAnimator = (props: Props) => {
+  const {
+    onAnimationEnd,
+    enterAnimationLocation,
+    exitAnimationLocation,
+    forwardedRef,
+    style: propsStyle,
+    onLayout: propsOnLayout,
+    ...others
+  } = props;
+  const containerRef = React.createRef<RNView>();
+  const {onLayout: hiddenLocationOnLayout, hiddenLocations} = useHiddenLocations({containerRef});
+  const visible = useSharedValue<boolean>(!enterAnimationLocation);
+
+  const getLocation = (location?: HiddenLocation) => {
```

Comment:
One more thing. 
We should aim using hooks to handle logic. 

Here, some logic is inside a hook while some is part of the component code. 

I wouldn't go crazy moving each single `if..else` to a hook. 
But in this case, wouldn't it be nice to have a general hook that handles these locations functionalities? 

Take a another look at the code, see if you can more relevant logic code under a hook.

Reply:
I don't think I see what you mean\see; would you like to talk about it or perhaps delay this to a future PR (`useHiddenLocations` will be used by `PanView` as well, so I might see some re-usable code).

Reply:
Why the hook will be used in PanView? 
I don't think we need to make this hook generic and multi purpose.. 

I think that as a thumb rule we should try using hooks to move out logic code out of the component code
While working on this component, try to focus only on it and avoid connecting it to the PanView

Reply:
It is needed so the animation is more precise (i.e. shorter); currently it's not dismissing to the correct location.